internal/controller/authenticator/jwtcachefiller/jwtcachefiller.go

```diff
@@ -217,21 +216,29 @@ func (c *jwtCacheFillerController) Sync(ctx controllerlib.Context) error {
 	errs = append(errs, jwksFetchErr)
 	okSoFar = okSoFar && jwksFetchErr == nil
 
-	// Make a deep copy of the spec so we aren't storing pointers to something that the informer cache
-	// may mutate! We don't store status as status is derived from spec.
-	cachedAuthenticator, conditions, err := c.newCachedJWTAuthenticator(
+	newJWTAuthenticatorForCache, conditions, err := c.newCachedJWTAuthenticator(
 		client,
-		obj.Spec.DeepCopy(),
+		obj.Spec.DeepCopy(), // deep copy to avoid caching original object
 		keySet,
 		conditions,
 		okSoFar)
 	errs = append(errs, err)
 
-	if !conditionsutil.HadErrorCondition(conditions) {
-		c.cache.Store(cacheKey, cachedAuthenticator)
-		c.log.Info("added new jwt authenticator", "jwtAuthenticator", klog.KObj(obj), "issuer", obj.Spec.Issuer)
+	if conditionsutil.HadErrorCondition(conditions) {
+		// The authenticator was determined to be invalid. Remove it from the cache, in case it was previously
+		// validated and cached. Do not allow an old, previously validated spec of the authenticator to continue
+		// being used for authentication.
+		c.cache.Delete(cacheKey)
+	} else {
+		c.cache.Store(cacheKey, newJWTAuthenticatorForCache)
+		c.log.WithValues("jwtAuthenticator", klog.KObj(obj), "issuer", obj.Spec.Issuer).
+			Info("added new jwt authenticator")
 	}
 
+	// In case we just overwrote or deleted the authenticator from the cache, clean up the old instance
+	// to avoid leaking goroutines. It's safe to call Close() on nil.
+	jwtAuthenticatorFromCache.Close()
```

Comment:
Could we move this call closer to where this variable is defined? We know that it will always be called if the specs are different.

Maybe this?

```golang
	var jwtAuthenticatorFromCache *cachedJWTAuthenticator
	if valueFromCache := c.cache.Get(cacheKey); valueFromCache != nil {
		jwtAuthenticatorFromCache = c.cacheValueAsJWTAuthenticator(valueFromCache)
		if jwtAuthenticatorFromCache != nil && reflect.DeepEqual(jwtAuthenticatorFromCache.spec, &obj.Spec) {
			c.log.WithValues("jwtAuthenticator", klog.KObj(obj), "issuer", obj.Spec.Issuer).
				Info("actual jwt authenticator and desired jwt authenticator are the same")
			// Stop, no more work to be done. This authenticator is already validated and cached.
			return nil
		}
	}
	// We will either delete or update the cache later, so close this now to avoid leaking goroutines.
	// It's safe to call Close() on nil.
	jwtAuthenticatorFromCache.Close()
```

Reply:
Closing it before it is removed from the in-memory cache could create a short period of time where users might not be able to authenticate. It's safer to close it after it is no longer in use.

Reply:
Will add a comment to the code to clarify.

Reply:
Updated in the second commit on this PR.

Reply:
got it